src/DatumCacheWsp.purs

```diff
@@ -0,0 +1,222 @@
+module DatumCacheWsp
+  ( DatumCacheResponse(..)
+  , DatumCacheRequest(..)
+  , DatumCacheMethod(..)
+  , JsonWspRequest
+  , JsonWspResponse
+  , WspFault(WspFault)
+  , faultToString
+  , jsonWspRequest
+  , parseJsonWspResponse
+  , responseMethod
+  , requestMethodName
+  ) where
+
+import Aeson
+  ( Aeson
+  , decodeAeson
+  , getNestedAeson
+  , toStringifiedNumbersJson
+  )
+import Control.Alt (map, (<$), (<$>), (<|>))
+import Control.Bind ((=<<), bind)
+import Control.Category ((<<<))
+import Data.Argonaut
+  ( Json
+  , JsonDecodeError(UnexpectedValue)
+  , decodeJson
+  , encodeJson
+  , jsonNull
+  , stringify
+  )
+import Data.Bifunctor (lmap)
+import Data.Either (Either(Left, Right), note)
+import Data.Eq (class Eq)
+import Data.Function ((>>>), const, ($))
+import Data.Maybe (Maybe(Just, Nothing), maybe)
+import Data.Newtype (unwrap)
+import Data.Traversable (traverse)
+import Data.Unit (Unit, unit)
+import Serialization.Address (BlockId, Slot)
+import Types.ByteArray (byteArrayToHex, hexToByteArray)
+import Types.PlutusData (PlutusData, DatumHash)
+import Types.Transaction (DataHash(DataHash))
+
+newtype WspFault = WspFault Json
+
+faultToString :: WspFault -> String
+faultToString (WspFault j) = stringify j
+
+type JsonWspRequest =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , args :: Json
+  }
+
+type JsonWspResponse =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , result :: Maybe Aeson
+  , fault :: Maybe Aeson
+  }
+
+data DatumCacheMethod
+  = GetDatumByHash
+  | GetDatumsByHashes
+  | StartFetchBlocks
+  | CancelFetchBlocks
+  | DatumFilterAddHashes
+  | DatumFilterRemoveHashes
+  | DatumFilterSetHashes
+  | DatumFilterGetHashes
+
+derive instance Eq DatumCacheMethod
+
+datumCacheMethodToString :: DatumCacheMethod -> String
```

Comment:
Well, it might produce just this output, but it is safer to use the exact values for the fromString function. So I could do it with a map lookup or an if clause, but this seems more straightforward and obvious.

Reply:
OK, in that case how about `show = datumCacheMethodToString`?

Reply:
Sure!